### PR TITLE
try GeometryBasics metadata approaches

### LINKED ISSUE #4: Handling miscellaneous data in geojson
For now GeoJSONTables can only read properties and geometries while other info viz. crs, bbox etc. is left out. It'd be nice to have some lower level functions to give users the option to do anything with it.

### LINKED ISSUE #5: Support for non tabular files
Currently the package only supports reading tabular geojson formats. We also need to have to add support for non tabular files. For instance a single point, or single feature, is also a valid GeoJSON, hence should be parsed accordingly.

### LINKED ISSUE #6: Missing properties
It is necessary to handle the cases where properties appear in one feature, but not other. What can be a possible work around for that case. Fill it up with `missing`?

### LINKED ISSUE #7: Bug in Polygon construction
So we manually give exterior and interiors to the `Polygon` method [here](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/geometry_basics.jl#L36-L49). 

We calculate `nring` and if `nring == 1`  there's only an exterior. But if `nring>1` , it implies that we have interiors too. So we create an array of interiors.

`Technically if we have nring = 3, it implies that we have one exterior and 2 interiors. `

The bug here is https://github.com/visr/GeoJSONTables.jl/blob/meta/src/geometry_basics.jl#L43 where we are allocating `nring` array size for interiors whereas it should be `nring-1`.

This leaves an extra `#undef` in the interiors array which makes it's way into the final GeometryBasics `Polygon`.

### PR DESCRIPTION
**Pull request by visr**
It's nice to use this as a basis for exploring the GeometryBasics approach to metadata, right now for https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49.

Since GeoJSON is a well defined standard, we know exactly what we need. Currently in this PR the switch to GeometryBasics geometries has already been made. To explore the needs for a future metadata in GeometryBasics, I avoided using meta geometries for know, and simply used `Feature` as a composed geometry and named tuple. `FeatureCollection` currently holds a vector of `Feature`s, but should be switched to StructArrays as suggested in https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49.

Supersedes #2

### CODE REVIEW

File: src/GeoJSONTables.jl [hunk 1]
```diff
@@ -18,7 +18,9 @@ function read(source)
     jsonfeatures = get(fc, :features, nothing)
     if get(fc, :type, nothing) == "FeatureCollection" && jsonfeatures isa JSON3.Array
         # TODO fill empty NamedTuple with properties
```
Comment (visr) [inline src/GeoJSONTables.jl]: TODO can be removed now

File: test/runtests.jl [hunk 2]
```diff
@@ -39,7 +39,7 @@ featurecollections = [g, multipolygon, realmultipolygon, polyline, point, pointn
 
         f1, _ = iterate(t)
         @test f1 isa GeoJSONTables.Feature
-        @test isempty(Base.propertynames(f1)) # .== [:cartodb_id, :addr1, :addr2, :park])
+        @test !isempty(Base.propertynames(f1)  .== [:cartodb_id, :addr1, :addr2, :park])
```
Comment (visr) [inline test/runtests.jl]: This doesn't seem right. You are checking if the comparison is empty.

Could you add some more tests? Now we only know it's not empty, but not that it is working. You could check the keys, values, types etc.

Comment (Sov-trotter) [reply test/runtests.jl]: Oh yeah sure. I had planned on adding more tests later on, for now I wanted to align to the tests you had added.

Comment (visr) [reply test/runtests.jl]: We probably redo the testing at some point yeah, but in general I think each new functionality should include at least some minimal tests, others you can assume in many cases that it doesn't work. At least that's how it works for me, most code I write and not test is broken.

**Comment by Sov-trotter** [comment]
The current implemenation is starting to hold well for homogeneous types. 
Wherein `GeoJSONTables.Feature{GeometryBasics.Point{2,Int64},(:a, :b),Tuple{String,Int64}}` comes out of `println(f)` in the `staticschema` method.
    
While for heterogeneous data `Feature{T,(:a, :b),Tuple{String,Int64}}` comes out  or `Feature{T,(:a, :b),Types} where Types where T`
So basically we're getting T/Types instead of concrete types depending on whatever is heterogeneous with a 
`MethodError: no method matching getnamestypes(::Type{Feature{T,(:a, :b),Types} where Types where T})`.

The this type inconsistency occurs while creating the vector of Feature structs -> [Line - 34](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/SA_test.jl#L34)
I tried to overload the `getnamestypes` method with `Any` to get it working, but that didn't help either
@piever Do you think I am missing something here?

**Comment by visr** [comment]
I don't fully understand the issue for heterogeneous types. We are using `Feature` with type parameters to have the ability to represent both homogeneous (`Feature{Point{2, Float64}}`) and heterogeneous (`Feature{Any/Union}`) feature collections. The `getnametypes` error seems to be a misspelled function name, it should be `getnamestypes`.

**Comment by Sov-trotter** [comment]
Oops! I misspelled `getnamestypes` in the above comment. I have fixed it now.

File: src/SA_test.jl [hunk 3]
```diff
@@ -0,0 +1,39 @@
+using GeometryBasics
+using GeometryBasics.StructArrays
+
+p1 = Point(2, 1)
+p2 = Point(3, 2)
+
+Feature(x; kwargs...) = Feature(x, values(kwargs))
+
+Base.getproperty(x::Feature, s::Symbol) = s == :geometry ? getfield(x, 1) : getproperty(getfield(x, 2), s) # looking for `s` in `properties`
```
Comment (piever) [inline src/SA_test.jl]: It's incorrect to define this without also overloading `propertynames`.

Comment (Sov-trotter) [reply src/SA_test.jl]: You mean `getnamestypes` should be replaced by `Base.propertynames`?

Comment (Sov-trotter) [reply src/SA_test.jl]: And a separate method for types ?

**Comment by piever** [comment]
With these types of custom types, you can't really expect StructArrays to know how to widen things correctly upon collection, or at least, I don't know a good way to do it in practice: how would StructArrays figure out that the first type parameter is linked to the first field? In principle, one could allow the widening mechanism to be customizable, but that is not implemented in StructArrays at the moment.

I think it's much easier to create the `StructArray` from a list of vectors. If you have a `Feature` iterator, you could use `Tables.columntable` (Tables is already a StructArrays dependency, so I guess it's OK to use it) on the iterator to change the storage to "column-based".

So something like

```julia
maketable(iter) = maketable(Tables.columntable(iter)::NamedTuple)

# assuming `geometry` is first in `propertynames(f::Feature)`
maketable(cols::NamedTuple) = maketable(first(cols), Base.tail(cols)) # you could also compute the types here with `Base.tuple_type_tail` and `Base.tuple_type_head`

function maketable(geometry, properties::NamedTuple{names, types}) where {names, types}
    F = Feature{eltype(geometry), names, StructArrays.eltypes(types)}
    return StructArray{F}(; geometry=geometry, properties...)
end
```

Note: this requires overload both `propertynames` and `getproperty` for `Feature`.

**Comment by Sov-trotter** [comment]
` In principle, one could allow the widening mechanism to be customizable, but that is not implemented in StructArrays at the moment.`

Not sure if it's correct, but I wonder if it was possible to write a `collect_structarrays` approach for the above method for non-standard data layouts?

File: src/SA_test.jl (hunk 3 above)
Comment (visr) [reply src/SA_test.jl]: `Base.propertynames` needs to be implemented for `Feature`, this is separate from the `getnamestypes` function, which is just a little helper function that is only used in the StructArrays README example. See the help of `propertynames` for what it should return.

**Comment by visr** [comment]
Let's keep it as simple as possible, and only focus on constructing it from a list of vectors for now.

**Comment by Sov-trotter** [comment]
The new approach works well!  For weird reasons I was getting a `type NamedTuple has no field properties` error. 
Unknowingly I put everything in the old environment(one with the `createinstance`, `staticschema` methods) and the error was gone!
I had it tested a few times, and things work well, so the latest commits include a `StructArrays.staticschema` method. I wonder what must be the reason for it? Maybe it was supposed to be written this way? Will test it a few more times before finally integrating everything?

**Comment by Sov-trotter** [comment]
Screenshot from 2020-07-09 22-28-07

**Comment by Sov-trotter** [comment]
Should I turn off the codecov comments?

**Comment by visr** [comment]
Yes please :)
Can be done with such a file: https://github.com/Deltares/Wflow.jl/blob/master/.codecov.yml

**Comment by Sov-trotter** [comment]
So currently [this](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L3-L5) is the way we're currently getting in the dependencies. Is there a better/clean way to write this ?

**Comment by visr** [comment]
I think that's fine, we can leave it like that.

**Comment by Sov-trotter** [comment]
Also the nomenclature might create some confusion [here](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L52) which returns geometries along with the names of properties present in the `Feature` struct, while this [here](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L58) will return only properties/metadata. 

So it will be weird to get geometry and metadata names out of `Base.propertynames()` and only metadata out of `properties()`

File: src/GeoJSONTables.jl [hunk 4]
```diff
@@ -101,10 +101,13 @@ function miss(x)
     end
 end
 
+Base.show(io::IO, fc::StructArray) = println(io, "FeatureCollection with $(length(fc)) Features")
```
Comment (visr) [inline src/GeoJSONTables.jl]: This overwrites the show method that is defined in StructArrays. We shouldn't do that.

Comment (Sov-trotter) [reply src/GeoJSONTables.jl]: Oh. I only found one here https://github.com/JuliaArrays/StructArrays.jl/blob/4483b57939f82828be51bc1e87c4a2de5f49ccf5/src/lazy.jl#L19 for LazyRow not StructArray so went ahead with it?

Comment (Sov-trotter) [reply src/GeoJSONTables.jl]: AH, found it https://github.com/JuliaArrays/StructArrays.jl/blob/4483b57939f82828be51bc1e87c4a2de5f49ccf5/src/structarray.jl#L223

Comment (visr) [reply src/GeoJSONTables.jl]: Hmm yeah you maybe right that it is not defined in StructArrays. But still it's not good practice to overload base functions on types you don't own.

File: .travis.yml [hunk 5]
```diff
@@ -1,7 +1,6 @@
 language: julia
 julia:
-  - 1.0
-  - 1.1
-  - 1.2
-  - 1.3
-  - nightly
+  - 1
+after_success:
```
Comment (visr) [inline .travis.yml]: can be removed

File: .travis.yml [hunk 6]
```diff
@@ -1,7 +1,6 @@
...
+codecov: true
+# - julia -e 'cd(Pkg.dir("GeoJSONTables")); Pkg.add("Coverage"); using Coverage; Codecov.submit(Codecov.process_folder())'
```
Comment (visr) [inline .travis.yml]: can be removed

File: README.md [hunk 7]
```diff
@@ -1,33 +1,61 @@
...
-Read [GeoJSON](https://geojson.org/) [FeatureCollections](https://tools.ietf.org/html/rfc7946#section-3.3) using [JSON3.jl](https://github.com/quinnj/JSON3.jl), and provide the [Tables.jl](https://github.com/JuliaData/Tables.jl) interface.
+Read [GeoJSON](https://geojson.org/) [FeatureCollections](https://tools.ietf.org/html/rfc7946#section-3.3) using [JSON3.jl](https://github.com/quinnj/JSON3.jl), and provide the [Tables.jl](https://github.com/JuliaData/Tables.jl) interface via [StructArrays.jl](https://github.com/JuliaArrays/StructArrays.jl).
 
-This package is unregistered and in development, so expect changes. It only supports reading, and only of FeatureCollections.
-
-This package is heavily inspired by, and borrows code from, [JSONTables.jl](https://github.com/JuliaData/JSONTables.jl), which
+This package is unregistered and in development, so expect changes. 
+For now it supports reading, of the following 
+* GeoJSON FeatureCollections
+* [GeometryBasics](https://github.com/JuliaGeometry/GeometryBasics.jl) geometries as Features
```
Comment (visr) [inline README.md:10]: It's best to not mix up geometries and features since in GeoJSON they are separate well defined things.

File: README.md [hunk 8]
```diff
@@ -1,33 +1,61 @@
...
-julia> using GeoJSONTables, DataFrames
+julia> using GeoJSONTables, DataFrames, GeometryBasics
 
 julia> jsonbytes = read("path/to/a.geojson");
 
+#= 
+   A true flag can be passed in the read function 
+   to read the jsonbytes as a JSON3 dict 
+   the default, false, will read it into a StructArray 
+=#
```
Comment (visr) [inline README.md]: I'm afraid that having a boolean flag like this is not really clear to users, and that it is better to offer it in a separate function altogether. Ah but if it just returns a JSON3 Object, people can just use the JSON3 function right? We could point to it in the docs for if people need lower level access though.

File: README.md [hunk 9]
```diff
@@ -1,33 +1,61 @@
...
+
 julia> first(fc)
-Feature with geometry type Polygon and properties Symbol[:geometry, :timestamp, :version, :changeset, :user, :uid, :area, :highway, :type, :id]
+Feature with geometry type Polygon and properties (:geometry, :timestamp, :version, :changeset, :user, :uid, :area, :highway, :type, :id)
 
 # use the Tables interface to convert the format, extract data, or iterate over the rows
 julia> df = DataFrame(fc)
+
+# GeometryBasics geometries can be passed along with metadata, into a Feature
+julia> f = GeoJSONTables.geo2feat(Point(1.0, 2.0), city = "Mumbai", rainfall = 1010)
```
Comment (visr) [inline README.md]: It looks like this could also be a constructor for `Feature`: `GeoJSONTables.Feature(Point(1.0, 2.0), city = "Mumbai", rainfall = 1010))` Does this already work?

File: src/GeoJSONTables.jl [hunk 10]
```diff
@@ -1,52 +1,73 @@
...
+    a = Symbol[]
     fc = JSON3.read(source)
-    features = get(fc, :features, nothing)
-    if get(fc, :type, nothing) == "FeatureCollection" && features isa JSON3.Array
-        FeatureCollection{typeof(features)}(features)
+    jsonfeatures = get(fc, :features, nothing)
+    if get(fc, :type, nothing) == "FeatureCollection" && jsonfeatures isa JSON3.Array
+        for f in jsonfeatures 
+            prop = f.properties
+            if !=(get(f, :properties, nothing), nothing)
```
Comment (visr) [inline src/GeoJSONTables.jl]: Nit: I prefer the inline `a != b`, think it is a bit clearer, see also below.

File: src/GeoJSONTables.jl [hunk 11]
```diff
@@ -1,52 +1,73 @@
...
+    else 
+        readtable(source)        
+    end
+end
+
+"""
+reads raw jsonbytes into a StructArray table
+"""
+function readtable(source)
+    features = Feature[]
```
Comment (visr) [inline src/GeoJSONTables.jl:16]: It would be more efficient to directly build vectors, one for the geometries, and one for each field, rather than a vector of features that will still end up as the above in memory later on when you create the structarray.

File: src/GeoJSONTables.jl [hunk 12]
```diff
@@ -55,33 +76,57 @@ Returns missing for null/nothing or not present, to work nicely with
...
-end
-
-@inline function Base.iterate(fc::FeatureCollection)
-    st = iterate(json(fc))
-    st === nothing && return nothing
-    val, state = st
-    return Feature(val), state
+function Base.getproperty(f::Feature, s::Symbol)
+    if s == :geometry
+        val = getfield(f, 1)
```
Comment (visr) [inline src/GeoJSONTables.jl]: Could you use the above defined `geometry` function here?

File: src/GeoJSONTables.jl [hunk 13]
```diff
@@ -55,33 +76,57 @@ Returns missing for null/nothing or not present, to work nicely with
...
+        miss(val)
+    else
+        val = getproperty(getfield(f, 2), s)
```
Comment (visr) [inline src/GeoJSONTables.jl]: Could you use the above defined `properties` function here?

File: .travis.yml (hunk 5 above)
Comment (Sov-trotter) [reply .travis.yml]: I am not sure what you're referring to here?

Comment (visr) [reply .travis.yml]: Just the line. It seems like `after_success:` can be removed since, it's an empty block?

File: README.md (hunk 8 above)
Comment (Sov-trotter) [reply README.md]: So it's better to remove this entirely?

File: README.md (hunk 9 above)
Comment (Sov-trotter) [reply README.md]: Yeah. https://github.com/visr/GeoJSONTables.jl/blob/meta/src/GeoJSONTables.jl#L118-L126

File: src/GeoJSONTables.jl (hunk 10 above)
Comment (Sov-trotter) [reply src/GeoJSONTables.jl]: Yeah. So vscode suggested me this. xD 
>New types should generally not implement this, and rely on the fallback definition !=(x,y) = !(x==y) instead.

File: .travis.yml (hunk 5 above)
Comment (Sov-trotter) [reply .travis.yml]: The block isn't empty I guess.` codecov:true` is inside it?

File: README.md (hunk 9 above)
Comment (Sov-trotter) [reply README.md]: So `geo2feat()` will handle it, even the `missing` cases nicely.

File: src/GeoJSONTables.jl (hunk 11 above)
Comment (Sov-trotter) [reply src/GeoJSONTables.jl:16]: But in that case we won't be able to construct the `Feature`? You're referring to this https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49#issuecomment-651203611 right?

File: .travis.yml (hunk 5 above)
Comment (visr) [reply .travis.yml]: That's not insided, because it is not indented.

File: README.md (hunk 9 above)
Comment (visr) [reply README.md]: I think the link is outdated now, not sure what lines you were referring to. It's good to press y in github before copying the link, that will put the commit hash in there such that the lines will stay the ones you mean.

But I don't understand your last comment. So `geo2feat()` does things that `Feature()` doesn't related to `missing`? What is that exactly? And can we put that behavior in the `Feature` constructor as well?

File: src/GeoJSONTables.jl (hunk 11 above)
Comment (visr) [reply src/GeoJSONTables.jl:16]: Yeah indeed that's what I mean. But Pietro has shown us how to construct the StructArray from vectors with Feature rows right, like here: https://github.com/JuliaGeometry/GeometryBasics.jl/issues/49#issuecomment-654120668?

File: src/GeoJSONTables.jl [hunk 14]
```diff
@@ -35,7 +35,7 @@ function readtable(source)
     if get(fc, :type, nothing) == "FeatureCollection" && jsonfeatures isa JSON3.Array
         for f in jsonfeatures 
             prop = f.properties
-            if !=(get(f, :properties, nothing), nothing)
+            if get(f, :properties, nothing) != nothing
```
Comment (visr) [inline src/GeoJSONTables.jl]: For `nothing` it can be faster to do `=== / !==` versus `== / !=`. You can also use `isnothing`. Note in the lines below there are still a few `!=(` cases as well.

File: src/GeoJSONTables.jl (hunk 11 above)
Comment (Sov-trotter) [reply src/GeoJSONTables.jl:16]: So going by this we will get properties as a StructArray too, hence nested, are we okay with it?

Comment (Sov-trotter) [reply src/GeoJSONTables.jl:16]: One thing we can do here is, test out both the implementations for performance and settle for the performant one?

File: README.md (hunk 9 above)
Comment (Sov-trotter) [reply README.md]: Okay. Thanks!
So the link was referring to `geo2feat` methods https://github.com/visr/GeoJSONTables.jl/blob/a8fc7bbfa61045dd1103ff9c5fd69a712b86a2fe/src/GeoJSONTables.jl#L105-L113

So what we want is, like you said earlier that even something like a `Point(1, 2)` is a valid geojson format and we should accept it as well. And I also wanted to be consistent with the currently existing `GeoJSON.jl` package. [This](https://github.com/visr/GeoJSONTables.jl/issues/5#issuecomment-659621092 ) comment includes the details. 

We can also use the existing `Feature` constructor as well. But I was concerned that it might run into conflicts with the actual usage of Feature [here](https://github.com/visr/GeoJSONTables.jl/blob/a8fc7bbfa61045dd1103ff9c5fd69a712b86a2fe/src/GeoJSONTables.jl#L34-L41) for which the [constructor](https://github.com/visr/GeoJSONTables.jl/blob/a8fc7bbfa61045dd1103ff9c5fd69a712b86a2fe/src/GeoJSONTables.jl#L12) already exists.

File: .travis.yml (hunk 5 above)
Comment (Sov-trotter) [reply .travis.yml]: Ah yeah!

File: README.md (hunk 9 above)
Comment (visr) [reply README.md]: > for which the constructor already exists

Luckily we have multiple dispatch

> Point(1, 2) is a valid geojson format and we should accept it as well

True, but that should just be a point and not a Feature. I don't see what `geo2feat` does in general, other than acting as a `Feature` constructor.

File: src/GeoJSONTables.jl (hunk 11 above)
Comment (visr) [reply src/GeoJSONTables.jl:16]: I'd prefer the non nested approach if we can get that to work.

Comment (Sov-trotter) [reply src/GeoJSONTables.jl:16]: Yeah, the [current](https://github.com/visr/GeoJSONTables.jl/blob/meta/src/Struct_Arrays.jl) approach is non-nested indeed. [This](https://github.com/JuliaArrays/StructArrays.jl#advanced-structures-with-non-standard-data-layout) is what @piever suggested with some [tweaking](https://github.com/visr/GeoJSONTables.jl/pull/3#issuecomment-656099490)

File: README.md (hunk 9 above)
Comment (Sov-trotter) [reply README.md:41]: Ah sure. I'll try it out then.

**Comment by visr** [comment]
Closing as we've decided to keep the wrapped JSON3 Objects as a kind of lazy geometry type, and implement GeoInterface on them, such that we can easily convert to GeometryBasics or other more useful geometry types.
